README.md
# Rust MAX44009/MAX44007 Ambient Light Sensor Driver

[![crates.io](https://img.shields.io/crates/v/max44009.svg)](https://crates.io/crates/max44009)
[![Docs](https://docs.rs/max44009/badge.svg)](https://docs.rs/max44009)
[![Build Status](https://github.com/eldruin/max44009-rs/workflows/Build/badge.svg)](https://github.com/eldruin/max44009-rs/actions?query=workflow%3ABuild)
[![Coverage Status](https://coveralls.io/repos/github/eldruin/max44009-rs/badge.svg?branch=master)](https://coveralls.io/github/eldruin/max44009-rs?branch=master)

This is a platform agnostic Rust driver for the MAX44009 and MAX44007 ambient
light sensors, using the [`embedded-hal`] traits.

This driver allows you to:
- Read lux measurement.
- Set the measurement mode.
- Set the configuration mode.
- Set the integration time.
- Set the current division ratio.
- Read the integration time.
- Read the current division ratio.
- Enable/disable interrupt generation.
- Check if an interrupt has happened.

## The devices
The MAX44009 and MAX44007 ambient light sensor feature an I2C digital output
that is ideal for a number of portable applications such as
smartphones, notebooks, and industrial sensors.
At less than 1μA operating current, the MAX44009 is the lowest power ambient
light sensor in the industry and features an ultra-wide 22-bit
dynamic range from 0.045 lux to 188,000 lux.
Low-light operation allows easy operation in dark-glass
applications.
The on-chip photodiode's spectral response is optimized to mimic
the human eye's perception of ambient light and incorporates
IR and UV blocking capability. The adaptive gain block
automatically selects the correct lux range to optimize the
counts/lux.

Datasheets: [MAX44007], [MAX44009]

[MAX44007]: https://datasheets.maximintegrated.com/en/ds/MAX44007.pdf
[MAX44009]: https://datasheets.maximintegrated.com/en/ds/MAX44009.pdf

## Usage

To use this driver, import this crate and an `embedded_hal` implementation,
then instantiate the appropriate device.

Please find additional examples using hardware in this repository: [driver-examples]

[driver-examples]: https://github.com/eldruin/driver-examples

```rust
use linux_embedded_hal::I2cdev;
use max44009::{Max44009, SlaveAddr};

fn main() {
    let dev = I2cdev::new("/dev/i2c-1").unwrap();
    let mut sensor = Max44009::new(dev, SlaveAddr::default());
    let lux = sensor.read_lux().unwrap();
    println!("lux: {}", lux);
}
```

## Support

For questions, issues, feature requests, and other changes, please file an
[issue in the github project](https://github.com/eldruin/max44009-rs/issues).

## License

Licensed under either of

 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or
   http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT)

at your option.

### Contributing

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall
be dual licensed as above, without any additional terms or conditions.

[`embedded-hal`]: https://github.com/rust-embedded/embedded-hal


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

...

## [0.2.0] - 2021-09-12

### Added
- Setting the configuration mode.
- Setting the integration time.
- Setting the current division ratio.
- Reading the integration time.
- Reading the current division ratio.
- Enabling/disabling interrupt generation.
- Checking if an interrupt has happened.
- Note about compatibility with MAX44007.

### Changed
- [breaking-change] Remove implementing `Default` for `Max44009`.
- Make types `ConfigurationMode`, `CurrentDivisionRatio`, `IntegrationTime`,
  `MeasurementMode` and `SlaveAddr` implement `Copy`.

## 0.1.0 - 2018-10-19

This is the initial release to crates.io. All changes will be documented in
this CHANGELOG.

[Unreleased]: https://github.com/eldruin/max44009-rs/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/eldruin/max44009-rs/compare/v0.1.0...v0.2.0



tests/tests.rs
extern crate max44009;
extern crate embedded_hal_mock as hal;
use max44009::{ Max44009, SlaveAddr, MeasurementMode, ConfigurationMode,
                IntegrationTime, CurrentDivisionRatio, Error };

const DEVICE_BASE_ADDRESS: u8 = 0b100_1010;

struct Register;

impl Register {
    const INT_STATUS        : u8 = 0x00;
    const INT_ENABLE        : u8 = 0x01;
    const CONFIGURATION     : u8 = 0x02;
    const LUX_HIGH          : u8 = 0x03;
    //const UPPER_THRESH_HIGH : u8 = 0x05;
    //const LOWER_THRESH_HIGH : u8 = 0x06;
    //const THRESH_TIMER      : u8 = 0x07;
}

fn setup<'a>(data: &'a[u8]) -> Max44009<hal::I2cMock<'a>> {
    let mut dev = hal::I2cMock::new();
    dev.set_read_data(&data);
    Max44009::new(dev, SlaveAddr::default())
}

fn check_sent_data(sensor: Max44009<hal::I2cMock>, data: &[u8]) {
    let dev = sensor.destroy();
    assert_eq!(dev.get_last_address(), Some(DEVICE_BASE_ADDRESS));
    assert_eq!(dev.get_write_data(), &data[..]);
}

fn assert_operation_not_available_error<T, E>(result: Result<T, Error<E>>) {
    match result {
        Err(Error::OperationNotAvailable) => (),
        _ => panic!("Did not return Error::OperationNotAvailable."),
    }
}

#[test]
fn can_read_interrupt_did_not_happened() {
    let mut dev = setup(&[0]);
    let interrupt_happened = dev.has_interrupt_happened().unwrap();
    assert!(!interrupt_happened);
    check_sent_data(dev, &[Register::INT_STATUS]);
}

#[test]
fn can_read_interrupt_happened() {
    let mut dev = setup(&[1]);
    let interrupt_happened = dev.has_interrupt_happened().unwrap();
    assert!(interrupt_happened);
    check_sent_data(dev, &[Register::INT_STATUS]);
}

#[test]
fn can_enable_interrupt() {
    let mut dev = setup(&[0]);
    dev.enable_interrupt().unwrap();
    check_sent_data(dev, &[Register::INT_ENABLE, 1]);
}

#[test]
fn can_disable_interrupt() {
    let mut dev = setup(&[0]);
    dev.disable_interrupt().unwrap();
    check_sent_data(dev, &[Register::INT_ENABLE, 0]);
}

#[test]
fn can_set_measurement_mode_once_every_800ms() {
    let mut dev = setup(&[0]);
    dev.set_measurement_mode(MeasurementMode::OnceEvery800ms).unwrap();
    check_sent_data(dev, &[Register::CONFIGURATION, 0]);
}

#[test]
fn can_set_measurement_mode_continuous() {
    let mut dev = setup(&[0]);
    dev.set_measurement_mode(MeasurementMode::Continuous).unwrap();
    check_sent_data(dev, &[Register::CONFIGURATION, 0b1000_0000]);
}

#[test]
fn can_read_lux() {
    let mut dev = setup(&[0, 1]);
    let lux = dev.read_lux().unwrap();
    assert_eq!(0.045, lux);
    check_sent_data(dev, &[Register::LUX_HIGH]);
}

#[test]
fn can_set_automatic_mode() {
    let mut dev = setup(&[0]);
    dev.set_configuration_mode(ConfigurationMode::Automatic).unwrap();
    check_sent_data(dev, &[Register::CONFIGURATION, 0b0000_0000]);
}

#[test]
fn can_set_manual_mode() {
    let mut dev = setup(&[0]);
    dev.set_configuration_mode(ConfigurationMode::Manual).unwrap();
    check_sent_data(dev, &[Register::CONFIGURATION, 0b0100_0000]);
}

#[test]
fn cannot_set_current_division_ratio_in_automatic_mode() {
    let mut dev = setup(&[0]);
    assert_operation_not_available_error(dev.set_current_division_ratio(CurrentDivisionRatio::One));
}

#[test]
fn cannot_set_integration_time_in_automatic_mode() {
    let mut dev = setup(&[0]);
    assert_operation_not_available_error(dev.set_integration_time(IntegrationTime::_100ms));
}

macro_rules! set_param_test {
    ($test_name:ident, $method:ident, $enum:ident::$variant:ident, $expected:expr) => {
        #[test]
        fn $test_name() {
            let mut dev = setup(&[0]);
            dev.set_configuration_mode(ConfigurationMode::Manual).unwrap();
            dev.$method($enum::$variant).unwrap();
            check_sent_data(dev, &[Register::CONFIGURATION, 0b0100_0000 | $expected]);
        }
    };
}

set_param_test!(can_set_cdr_one,        set_current_division_ratio, CurrentDivisionRatio::One,        0);
set_param_test!(can_set_cdr_one_eighth, set_current_division_ratio, CurrentDivisionRatio::OneEighth,  8);

set_param_test!(can_set_it_800ms,  set_integration_time, IntegrationTime::_800ms,  0);
set_param_test!(can_set_it_400ms,  set_integration_time, IntegrationTime::_400ms,  1);
set_param_test!(can_set_it_200ms,  set_integration_time, IntegrationTime::_200ms,  2);
set_param_test!(can_set_it_100ms,  set_integration_time, IntegrationTime::_100ms,  3);
set_param_test!(can_set_it_50ms,   set_integration_time, IntegrationTime::_50ms,   4);
set_param_test!(can_set_it_25ms,   set_integration_time, IntegrationTime::_25ms,   5);
set_param_test!(can_set_it_12_5ms, set_integration_time, IntegrationTime::_12_5ms, 6);
set_param_test!(can_set_it_6_25ms, set_integration_time, IntegrationTime::_6_25ms, 7);

macro_rules! read_param_test {
    ($test_name:ident, $method:ident, $input_data:expr, $enum:ident::$expected_variant:ident) => {
        #[test]
        fn $test_name() {
            let mut dev = setup(&[$input_data]);
            let it = dev.$method().unwrap();
            assert_eq!($enum::$expected_variant, it);
            check_sent_data(dev, &[Register::CONFIGURATION]);
        }
    }
}

read_param_test!(can_read_cdr_one,        read_current_division_ratio, 0, CurrentDivisionRatio::One      );
read_param_test!(can_read_cdr_one_eighth, read_current_division_ratio, 8, CurrentDivisionRatio::OneEighth);

read_param_test!(can_read_it_800ms,  read_integration_time, 0, IntegrationTime::_800ms);
read_param_test!(can_read_it_400ms,  read_integration_time, 1, IntegrationTime::_400ms);
read_param_test!(can_read_it_200ms,  read_integration_time, 2, IntegrationTime::_200ms);
read_param_test!(can_read_it_100ms,  read_integration_time, 3, IntegrationTime::_100ms);
read_param_test!(can_read_it_50ms,   read_integration_time, 4, IntegrationTime::_50ms);
read_param_test!(can_read_it_25ms,   read_integration_time, 5, IntegrationTime::_25ms);
read_param_test!(can_read_it_12_5ms, read_integration_time, 6, IntegrationTime::_12_5ms);
read_param_test!(can_read_it_6_25ms, read_integration_time, 7, IntegrationTime::_6_25ms);



tests/reading.rs
use embedded_hal_mock::i2c::Transaction as I2cTrans;
use max44009::{CurrentDivisionRatio as CDR, IntegrationTime as IT};
mod common;
use crate::common::{destroy, new, Register, DEV_BASE_ADDR};

#[test]
fn can_read_interrupt_did_not_happened() {
    let mut dev = new(&[I2cTrans::write_read(
        DEV_BASE_ADDR,
        vec![Register::INT_STATUS],
        vec![0],
    )]);
    let interrupt_happened = dev.has_interrupt_happened().unwrap();
    assert!(!interrupt_happened);
    destroy(dev);
}

#[test]
fn can_read_interrupt_happened() {
    let mut dev = new(&[I2cTrans::write_read(
        DEV_BASE_ADDR,
        vec![Register::INT_STATUS],
        vec![1],
    )]);
    let interrupt_happened = dev.has_interrupt_happened().unwrap();
    assert!(interrupt_happened);
    destroy(dev);
}

#[test]
fn can_read_lux() {
    let mut dev = new(&[I2cTrans::write_read(
        DEV_BASE_ADDR,
        vec![Register::LUX_HIGH],
        vec![0, 1],
    )]);
    let lux = dev.read_lux().unwrap();
    assert!((lux - 0.045).abs() < 0.001);
    destroy(dev);
}

macro_rules! read_param_test {
    ($test_name:ident, $method:ident, $input_data:expr, $enum:ident::$expected_variant:ident) => {
        #[test]
        fn $test_name() {
            let mut dev = new(&[I2cTrans::write_read(
                DEV_BASE_ADDR,
                vec![Register::CONFIGURATION],
                vec![$input_data],
            )]);
            let it = dev.$method().unwrap();
            assert_eq!($enum::$expected_variant, it);
            destroy(dev);
        }
    };
}

read_param_test!(can_read_cdr_one, read_current_division_ratio, 0, CDR::One);
read_param_test!(
    can_read_cdr_one_eighth,
    read_current_division_ratio,
    8,
    CDR::OneEighth
);

read_param_test!(can_read_it_800ms, read_integration_time, 0, IT::_800ms);
read_param_test!(can_read_it_400ms, read_integration_time, 1, IT::_400ms);
read_param_test!(can_read_it_200ms, read_integration_time, 2, IT::_200ms);
read_param_test!(can_read_it_100ms, read_integration_time, 3, IT::_100ms);
read_param_test!(can_read_it_50ms, read_integration_time, 4, IT::_50ms);
read_param_test!(can_read_it_25ms, read_integration_time, 5, IT::_25ms);
read_param_test!(can_read_it_12_5ms, read_integration_time, 6, IT::_12_5ms);
read_param_test!(can_read_it_6_25ms, read_integration_time, 7, IT::_6_25ms);


tests/common/mod.rs
use embedded_hal_mock::i2c::{Mock as I2cMock, Transaction as I2cTrans};
use max44009::{Max44009, SlaveAddr};

pub const DEV_BASE_ADDR: u8 = 0b100_1010;

pub struct Register;

#[allow(unused)]
impl Register {
    pub const INT_STATUS: u8 = 0x00;
    pub const INT_ENABLE: u8 = 0x01;
    pub const CONFIGURATION: u8 = 0x02;
    pub const LUX_HIGH: u8 = 0x03;
}

pub fn new(transactions: &[I2cTrans]) -> Max44009<I2cMock> {
    Max44009::new(I2cMock::new(transactions), SlaveAddr::default())
}

pub fn destroy(dev: Max44009<I2cMock>) {
    dev.destroy().done();
}


src/types.rs
//! Public types

/// All possible errors in this crate
#[derive(Debug)]
pub enum Error<E> {
    /// I²C bus error.
    I2C(E),
    /// A manual-configuration-mode-only was attempted while in automatic
    /// configuration mode.
    OperationNotAvailable,
}

/// Measurement mode
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum MeasurementMode {
    /// Once every 800ms mode (default).
    ///
    /// Measures lux intensity every 800ms regardless of the integration time.
    /// Sensor operates on lowest possible supply current.
    OnceEvery800ms,
    /// Continuous mode.
    ///
    /// Continuously measures lux intensity. As soon as a reading finishes,
    /// the next one begins. The actual cadence depends on the integration
    /// time selected.
    Continuous,
}

/// Configuration mode
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ConfigurationMode {
    /// Automatic mode (default).
    ///
    /// On-chip algorithm selects the integration time (100ms - 800ms) and
    /// the current division ratio
    Automatic,
    /// Manual mode.
    ///
    /// The user can select the integration time and the current division
    /// ratio manually.
    Manual,
}

/// Integration time
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum IntegrationTime {
    /// 6.25ms. (Only in manual mode)
    _6_25ms,
    /// 12.5ms. (Only in manual mode)
    _12_5ms,
    /// 25ms. (Only in manual mode)
    _25ms,
    /// 50ms. (Only in manual mode)
    _50ms,
    /// 100ms. (Preferred mode for high-brightness applications)
    _100ms,
    /// 200ms
    _200ms,
    /// 400ms
    _400ms,
    /// 800ms. (Preferred mode for boosting low-light sensitivity)
    _800ms,
}

/// Current division ratio
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum CurrentDivisionRatio {
    /// No current division (default).
    ///
    /// All the photodiode current goes to the ADC.
    One,
    /// 1/8 current division ratio.
    ///
    /// Only 1/8 of the photodiode current goes to the ADC. This mode is used in
    /// high-brightness situations.
    OneEighth,
}

/// Possible slave addresses
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum SlaveAddr {
    /// Default slave address
    Default,
    /// Alternative slave address providing bit value for A0
    Alternative(bool),
}

impl Default for SlaveAddr {
    /// Default slave address
    fn default() -> Self {
        SlaveAddr::Default
    }
}

impl SlaveAddr {
    pub(crate) fn addr(self, default: u8) -> u8 {
        match self {
            SlaveAddr::Default => default,
            SlaveAddr::Alternative(a0) => default | a0 as u8,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::DEVICE_BASE_ADDRESS;

    #[test]
    fn can_get_default_address() {
        let addr = SlaveAddr::default();
        assert_eq!(DEVICE_BASE_ADDRESS, addr.addr(DEVICE_BASE_ADDRESS));
    }

    #[test]
    fn can_generate_alternative_addresses() {
        assert_eq!(
            0b100_1010,
            SlaveAddr::Alternative(false).addr(DEVICE_BASE_ADDRESS)
        );
        assert_eq!(
            0b100_1011,
            SlaveAddr::Alternative(true).addr(DEVICE_BASE_ADDRESS)
        );
    }
}


src/reading.rs
use crate::{BitFlags, CurrentDivisionRatio, Error, IntegrationTime, Max44009, Register};
use embedded_hal::blocking::i2c;

impl<I2C, E> Max44009<I2C>
where
    I2C: i2c::WriteRead<Error = E>,
{
    /// Reads whether an interrupt has happened.
    pub fn has_interrupt_happened(&mut self) -> Result<bool, Error<E>> {
        let mut data = [0];
        self.i2c
            .write_read(self.address, &[Register::INT_STATUS], &mut data)
            .map_err(Error::I2C)
            .and(Ok(data[0] != 0))
    }

    /// Read the lux intensity.
    pub fn read_lux(&mut self) -> Result<f32, Error<E>> {
        let mut data = [0; 2];
        self.i2c
            .write_read(self.address, &[Register::LUX_HIGH], &mut data)
            .map_err(Error::I2C)
            .and(Ok(convert_to_lux(data[0], data[1])))
    }

    /// Read the integration time.
    pub fn read_integration_time(&mut self) -> Result<IntegrationTime, Error<E>> {
        let mut config = [0];
        self.i2c
            .write_read(self.address, &[Register::CONFIGURATION], &mut config)
            .map_err(Error::I2C)?;
        match config[0] & 0b0000_0111 {
            0 => Ok(IntegrationTime::_800ms),
            1 => Ok(IntegrationTime::_400ms),
            2 => Ok(IntegrationTime::_200ms),
            3 => Ok(IntegrationTime::_100ms),
            4 => Ok(IntegrationTime::_50ms),
            5 => Ok(IntegrationTime::_25ms),
            6 => Ok(IntegrationTime::_12_5ms),
            7 => Ok(IntegrationTime::_6_25ms),
            _ => panic!("Programming error!"),
        }
    }

    /// Read the current division ratio.
    pub fn read_current_division_ratio(&mut self) -> Result<CurrentDivisionRatio, Error<E>> {
        let mut config = [0];
        self.i2c
            .write_read(self.address, &[Register::CONFIGURATION], &mut config)
            .map_err(Error::I2C)?;
        if (config[0] & BitFlags::CDR) == 0 {
            Ok(CurrentDivisionRatio::One)
        } else {
            Ok(CurrentDivisionRatio::OneEighth)
        }
    }
}

fn convert_to_lux(msb: u8, lsb: u8) -> f32 {
    let mantissa = (msb & 0x0F) << 4 | (lsb & 0x0F);
    let exp = (msb & 0xF0) >> 4;
    (((1_u32) << exp) * u32::from(mantissa)) as f32 * 0.045
}

#[cfg(test)]
mod tests {
    use super::*;

    fn assert_near(a: f32, b: f32, epsilon: f32) {
        assert!((a - b).abs() < epsilon);
    }

    #[test]
    fn can_convert_to_lux() {
        assert_near(0.045, convert_to_lux(0b0000_0000, 0b0000_0001), 0.001);
        assert_near(0.72, convert_to_lux(0b0000_0001, 0b0000_0000), 0.001);
        assert_near(1.53, convert_to_lux(0b0001_0001, 0b0000_0001), 0.001);
        assert_near(188_006.0, convert_to_lux(0b1110_1111, 0b0000_1111), 0.5);
        assert_near(187_269.0, convert_to_lux(0b1110_1111, 0b0000_1110), 0.5);
        assert_near(176_947.0, convert_to_lux(0b1110_1111, 0b0000_0000), 0.5);
        assert_near(165_151.0, convert_to_lux(0b1110_1110, 0b0000_0000), 0.5);
    }
}


src/configuration.rs
use crate::{
    BitFlags, ConfigurationMode, CurrentDivisionRatio, Error, IntegrationTime, Max44009,
    MeasurementMode, Register,
};
use embedded_hal::blocking::i2c;

impl<I2C, E> Max44009<I2C>
where
    I2C: i2c::Write<Error = E>,
{
    /// Enable interrupt.
    ///
    /// The INT pin will be pulled low if the interrupt condition is triggered.
    pub fn enable_interrupt(&mut self) -> Result<(), Error<E>> {
        self.i2c
            .write(self.address, &[Register::INT_ENABLE, 1])
            .map_err(Error::I2C)
    }

    /// Disable interrupt.
    pub fn disable_interrupt(&mut self) -> Result<(), Error<E>> {
        self.i2c
            .write(self.address, &[Register::INT_ENABLE, 0])
            .map_err(Error::I2C)
    }

    /// Set the measurement mode.
    pub fn set_measurement_mode(&mut self, mode: MeasurementMode) -> Result<(), Error<E>> {
        let config = self.config;
        match mode {
            MeasurementMode::OnceEvery800ms => self.write_config(config & !BitFlags::CONTINUOUS),
            MeasurementMode::Continuous => self.write_config(config | BitFlags::CONTINUOUS),
        }
    }

    /// Set configuration mode.
    pub fn set_configuration_mode(&mut self, mode: ConfigurationMode) -> Result<(), Error<E>> {
        let config = self.config;
        match mode {
            ConfigurationMode::Automatic => self.write_config(config & !BitFlags::MANUAL),
            ConfigurationMode::Manual => self.write_config(config | BitFlags::MANUAL),
        }
    }

    /// Set integration time. (Only in manual configuration mode).
    pub fn set_integration_time(&mut self, it: IntegrationTime) -> Result<(), Error<E>> {
        self.assert_is_in_manual_mode()?;
        let config = self.config & 0b1111_1000;
        match it {
            IntegrationTime::_800ms => self.write_config(config),
            IntegrationTime::_400ms => self.write_config(config | 0x01),
            IntegrationTime::_200ms => self.write_config(config | 0x02),
            IntegrationTime::_100ms => self.write_config(config | 0x03),
            IntegrationTime::_50ms => self.write_config(config | 0x04),
            IntegrationTime::_25ms => self.write_config(config | 0x05),
            IntegrationTime::_12_5ms => self.write_config(config | 0x06),
            IntegrationTime::_6_25ms => self.write_config(config | 0x07),
        }
    }

    /// Set current division ratio. (Only in manual configuration mode).
    pub fn set_current_division_ratio(
        &mut self,
        cdr: CurrentDivisionRatio,
    ) -> Result<(), Error<E>> {
        self.assert_is_in_manual_mode()?;
        let config = self.config;
        match cdr {
            CurrentDivisionRatio::One => self.write_config(config & !BitFlags::CDR),
            CurrentDivisionRatio::OneEighth => self.write_config(config | BitFlags::CDR),
        }
    }

    fn write_config(&mut self, config: u8) -> Result<(), Error<E>> {
        self.i2c
            .write(self.address, &[Register::CONFIGURATION, config])
            .map_err(Error::I2C)?;
        self.config = config;
        Ok(())
    }

    fn assert_is_in_manual_mode(&self) -> Result<(), Error<E>> {
        if (self.config & BitFlags::MANUAL) == 0 {
            return Err(Error::OperationNotAvailable);
        }
        Ok(())
    }
}


src/lib.rs
//! This is a platform agnostic Rust driver for the MAX44009 and MAX44007 ambient
//! light sensors (ALS), based on the [`embedded-hal`] traits.
//!
//! [`embedded-hal`]: https://github.com/rust-embedded/embedded-hal
//!
//! This driver allows you to:
//! - Read lux measurement.
//! - Set the measurement mode.
//! - Set the configuration mode.
//! - Set the integration time.
//! - Set the current division ratio.
//! - Read the integration time.
//! - Read the current division ratio.
//! - Enable/disable interrupt generation.
//! - Check if an interrupt has happened.
//!
//! ## The devices
//! The MAX44009 and MAX44007 ambient light sensors feature an I2C digital output
//! that is ideal for a number of portable applications such as
//! smartphones, notebooks, and industrial sensors.
//! At less than 1μA operating current, the MAX44009 is the lowest power ambient
//! light sensor in the industry and features an ultra-wide 22-bit
//! dynamic range from 0.045 lux to 188,000 lux.
//! Low-light operation allows easy operation in dark-glass
//! applications.
//! The on-chip photodiode's spectral response is optimized to mimic
//! the human eye's perception of ambient light and incorporates
//! IR and UV blocking capability. The adaptive gain block
//! automatically selects the correct lux range to optimize the
//! counts/lux.
//!
//! Datasheets: [MAX44007], [MAX44009]
//!
//! [MAX44007]: https://datasheets.maximintegrated.com/en/ds/MAX44007.pdf
//! [MAX44009]: https://datasheets.maximintegrated.com/en/ds/MAX44009.pdf
//!
//! ## Usage examples (see also examples folder)
//!
//! To use this driver, import this crate and an `embedded_hal` implementation,
//! then instantiate the device.
//!
//! Please find additional examples using hardware in this repository: [driver-examples]
//!
//! [driver-examples]: https://github.com/eldruin/driver-examples
//!
//! ### Read lux
//!
//! Import this crate and an `embedded_hal` implementation, then instantiate
//! the device:
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use max44009::{ Max44009, SlaveAddr };
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let address = SlaveAddr::default();
//! let mut sensor = Max44009::new(dev, address);
//! let lux = sensor.read_lux().unwrap();
//! ```
//!
//! ### Provide an alternative address
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use max44009::{ Max44009, SlaveAddr };
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let a0 = true;
//! let address = SlaveAddr::Alternative(a0);
//! let mut sensor = Max44009::new(dev, address);
//! ```
//!
//! ### Enable interruptions and see if one has happened
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use max44009::{ Max44009, SlaveAddr };
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Max44009::new(dev, SlaveAddr::default());
//! sensor.enable_interrupt().unwrap();
//! if sensor.has_interrupt_happened().unwrap() {
//!     println!("Interrupt happened.");
//! }
//! ```
//!
//! ### Set the measurement mode to continuous
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use max44009::{ Max44009, SlaveAddr, MeasurementMode };
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Max44009::new(dev, SlaveAddr::default());
//! sensor.set_measurement_mode(MeasurementMode::Continuous).unwrap();
//! ```
//!
//! ### Read the parameters selected in automatic mode
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use max44009::{ Max44009, SlaveAddr };
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Max44009::new(dev, SlaveAddr::default());
//! let it  = sensor.read_integration_time().unwrap();
//! let cdr = sensor.read_current_division_ratio().unwrap();
//! ```
//!
//! ### Configure manually
//! - Set configuration mode to manual.
//! - Set current division ratio to 1/8.
//! - Set integration time to 100ms.
//!
//! ```no_run
//! use linux_embedded_hal::I2cdev;
//! use max44009::{ Max44009, SlaveAddr, ConfigurationMode,
//!                 CurrentDivisionRatio, IntegrationTime };
//!
//! let dev = I2cdev::new("/dev/i2c-1").unwrap();
//! let mut sensor = Max44009::new(dev, SlaveAddr::default());
//! sensor.set_configuration_mode(ConfigurationMode::Manual).unwrap();
//! sensor.set_current_division_ratio(CurrentDivisionRatio::OneEighth).unwrap();
//! sensor.set_integration_time(IntegrationTime::_100ms).unwrap();
//! ```
//!

#![doc(html_root_url = "https://docs.rs/max44009/0.2.0")]
#![deny(unsafe_code)]
#![deny(missing_docs)]
#![no_std]

use embedded_hal::blocking::i2c;

const DEVICE_BASE_ADDRESS: u8 = 0b100_1010;

struct Register;

impl Register {
    const INT_STATUS: u8 = 0x00;
    const INT_ENABLE: u8 = 0x01;
    const CONFIGURATION: u8 = 0x02;
    const LUX_HIGH: u8 = 0x03;
}

struct BitFlags;

impl BitFlags {
    const CONTINUOUS: u8 = 0b1000_0000;
    const MANUAL: u8 = 0b0100_0000;
    const CDR: u8 = 0b0000_1000;
}

/// MAX44009 ambient light sensor driver.
#[derive(Debug)]
pub struct Max44009<I2C> {
    /// The concrete I²C device implementation.
    i2c: I2C,
    /// The I²C device address.
    address: u8,
    /// Configuration register status.
    config: u8,
}

mod configuration;
mod reading;
mod types;
pub use crate::types::{
    ConfigurationMode, CurrentDivisionRatio, Error, IntegrationTime, MeasurementMode, SlaveAddr,
};

impl<I2C, E> Max44009<I2C>
where
    I2C: i2c::Write<Error = E>,
{
    /// Create new instance of the Max44009 device.
    pub fn new(i2c: I2C, address: SlaveAddr) -> Self {
        Max44009 {
            i2c,
            address: address.addr(DEVICE_BASE_ADDRESS),
            config: 0,
        }
    }

    /// Destroy driver instance, return I²C bus instance.
    pub fn destroy(self) -> I2C {
        self.i2c
    }
}
